### PR TITLE
Scheduler Prioirty 

### LINKED ISSUE #96: Add realtime scheduling priority to all threads
The linux kernel is not designed for high speed and responsive audio processing tasks. The default scheduling priorities leaves our threads running too infrequently.

Success criteria:
- Increase scheduling criteria of threads

### PR DESCRIPTION
**Pull request by bear-in-the-air**
closes #96

### CODE REVIEW

File: include/scheduler_helper.hh [hunk 1]
```diff
@@ -0,0 +1,12 @@
+#ifndef SCHEDULER_HELPER_HH
+#define SCHEDULER_HELPER_HH
+
+#include <cstring>
+#include <iostream>
+#include <pthread.h>
+class Scheduler_helper {
```
Comment (Jambls) [inline include/scheduler_helper.hh:7]: Can you document this to javadoc spec please xox

**Review by Jambls — COMMENTED** [review]
Looks good other than documentation.

**Review by Jambls — APPROVED** [review]

File: src/scheduler_helper.cc [hunk 2]
```diff
@@ -0,0 +1,17 @@
+#include "../include/scheduler_helper.hh"
+
+void Scheduler_helper::set_thread_priority(pthread_t handle, std::string const &threadname) {
+    // NOLINTNEXTLINE(cppcoreguidelines-pro-type-member-init) variable is handed in to function to be filled out
+    sched_param sch;
+    // NOLINTNEXTLINE(cppcoreguidelines-init-variables) variable is handed in to function to be filled out
+    int policy;
```
Comment (MHCooke) [inline src/scheduler_helper.cc:7]: Have you tried initialising these to be empty?

File: src/scheduler_helper.cc [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+void Scheduler_helper::set_thread_priority(pthread_t handle, std::string const &threadname) {
+
+    sched_param sch{0};
+    int policy = 0;
+    pthread_getschedparam(handle, &policy, &sch);
+    // NOLINTNEXTLINE(cppcoreguidelines-avoid-magic-numbers)
+    sch.sched_priority = 20;
+    if (pthread_setschedparam(handle, SCHED_FIFO, &sch)) {
+        // NOLINTNEXTLINE(concurrency-mt-unsafe)
+        std::cerr << "Failed to set sched param: " << std::strerror(errno) << std::endl;
```
Comment (MHCooke) [inline src/scheduler_helper.cc:12]: This isn't a meaningful error message

Comment (MarkAHarley) [reply src/scheduler_helper.cc:12]: Can't think of any way to get more information when this is seperated though.

Comment (bear-in-the-air) [reply src/scheduler_helper.cc]: Added the threadname to to the error output. 

I am aware that we will have arrays of some of the same threads, so would giving each thread a number be beneficial?

Comment (MHCooke) [reply src/scheduler_helper.cc]: I was thinking more like. "Failed to change priority for thread xxxx". The user doesn't care about sched params.

Comment (MarkAHarley) [reply src/scheduler_helper.cc]: that seems sensible to me.

**Review by MarkAHarley — APPROVED** [review]
Wait for Matthew to resolve his issue.

**Review by MarkAHarley — CHANGES_REQUESTED** [review]
Need to add priority stuff for alsa_out.

**Pull request merged**